### PR TITLE
[BEAM-11267] Remove unnecessary reshuffle for stateful ParDo after key…

### PR DESCRIPTION
**Pull request by dmvk**
https://issues.apache.org/jira/browse/BEAM-11267

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTranslationContext.java [hunk 1]
```diff
@@ -84,6 +85,17 @@ public void setOutputDataStream(PValue value, DataStream<?> set) {
     }
   }
 
+  <T extends PValue> void setProducer(T value, PTransform<?, T> producer) {
+    if (!producers.containsKey(value)) {
```
Comment (je-ik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTranslationContext.java:89]: Should this be turned to `Precondition.checkState`?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java [hunk 2]
```diff
@@ -49,6 +52,6 @@ public ByteBuffer getKey(WindowedValue<SingletonKeyedWorkItem<K, V>> value) thro
 
   @Override
   public TypeInformation<ByteBuffer> getProducedType() {
-    return new GenericTypeInfo<>(ByteBuffer.class);
+    return new CoderTypeInformation<>(FlinkKeyUtils.ByteBufferCoder.of(), pipelineOptions.get());
```
Comment (dmvk) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java:55]: Need to check Kryo internals whether this is breaking change 🤔

Comment (je-ik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java:55]: This looks unrelated, can you just explain this modification? I suppose it is correct, just wonder why it was `GenericTypeInfo` before.

**Review by je-ik — CHANGES_REQUESTED** [review]
+1 overall.
I have a concern about simply ignoring the fact that two producers register the output. We should take that as an error or handle that somehow.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java (hunk 2 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java:55]: Not sure if this is necessary. I wanted to ensure that the new "reinterpreted partitioning" is compatible with the one used by GBK / Combine.

The idea was if partitioning is not compatible, it may result in some state partitioning related glitches (eg. you wouldn't have local state for a key-group you need).

Second thoughts, flink selects target partition based on "pojo hash code" (not based on binary representation), so the previous version was probably compatible enough 🤔 

@mxm WDYT?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTranslationContext.java (hunk 1 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTranslationContext.java:89]: 👍 makes sense

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java (hunk 2 above)
Comment (je-ik) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java:55]: I would think that the partitioning is ensured by the `reinterpretAsKeyedStream` and will therefore be preserved from the previous shuffle phase. Is this not enough?

**Comment by je-ik** [comment]
Do we have a test (in flink runner) for the GBK -> stateful pardo pair? Not sure if there is one in ValidatesRunner suite.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java [hunk 3]
```diff
@@ -971,7 +987,9 @@ public void translateNode(
               .transform(fullName, outputTypeInfo, (OneInputStreamOperator) doFnOperator)
               .uid(fullName);
 
-      context.setOutputDataStream(context.getOutput(transform), outDataStream);
+      final PCollection<KV<K, Iterable<InputT>>> output = context.getOutput(transform);
+      context.setOutputDataStream(output, outDataStream);
+      context.setProducer(output, transform);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java:992]: It would be nice if these explicit calls wouldn't be required. I believe `context.setOutputDataStream` internally has the current transform available. So we could perform it internally in the context.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java [hunk 4]
```diff
@@ -1127,7 +1148,9 @@ public void translateNode(
 
         keyedWorkItemStream.getExecutionEnvironment().addOperator(rawFlinkTransform);
 
-        context.setOutputDataStream(context.getOutput(transform), outDataStream);
+        final PCollection<KV<K, OutputT>> output = context.getOutput(transform);
+        context.setOutputDataStream(output, outDataStream);
+        context.setProducer(output, transform);
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java:1153]: Same here.

**Review by mxm — COMMENTED** [review]
That's a decent optimization. Generally, looks good to me, apart from the comments. I'd appreciate if we could add a test which ensures that the optimization gets applied.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java (hunk 3 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java:992]: nice catch 👍

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java (hunk 2 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/WorkItemKeySelector.java:55]: I think that `reinterpretAsKeyedStream` just trusts the user and there are no safety nets in place.

**Comment by dmvk** [comment]
Run Flink ValidatesRunner

**Comment by dmvk** [comment]
@je-ik @mxm thanks for the review ;) I've incorporated your suggestions and it's ready for another pass

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java [hunk 5]
```diff
@@ -120,4 +138,95 @@ public void testAutoBalanceShardKeyCacheMaxSize() throws Exception {
...
+    final JobGraph stablePartitioning = getStatefulParDoAfterCombineChainingJobGraph(true);
+    final JobGraph unstablePartitioning = getStatefulParDoAfterCombineChainingJobGraph(false);
+    // We expect an extra shuffle stage for unstable partitioning.
+    Assert.assertEquals(
+        1,
+        Iterables.size(unstablePartitioning.getVertices())
+            - Iterables.size(stablePartitioning.getVertices()));
+  }
+
+  private JobGraph getStatefulParDoAfterCombineChainingJobGraph(boolean stablePartitioning) {
```
Comment (je-ik) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:153]: I would call this parameter `forcesShuffle` or something similar.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java [hunk 6]
```diff
@@ -120,4 +138,95 @@ public void testAutoBalanceShardKeyCacheMaxSize() throws Exception {
...
+    pipelineOptions.setRunner(FlinkRunner.class);
+    final Pipeline pipeline = Pipeline.create(pipelineOptions);
+    PCollection<KV<String, Long>> aggregate =
+        pipeline
+            .apply(Create.of("foo", "bar").withCoder(StringUtf8Coder.of()))
+            .apply(Count.perElement());
+    if (!stablePartitioning) {
+      // When we insert any element-wise "map" operation between aggregation and stateful ParDo, we
+      // can no longer assume that partitioning did not change, therefore we need an extra shuffle
+      aggregate = aggregate.apply(ParDo.of(new StatelessIdentityDoFn<>()));
```
Comment (je-ik) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:167]: Instead of the `StatelessIdentityDoFn` we could use `MapElements.into(...).via(e -> KV.of("", e.getValue())`, which would enforce shuffle semantically. That might improve readability a bit.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java [hunk 7]
```diff
@@ -120,4 +138,95 @@ public void testAutoBalanceShardKeyCacheMaxSize() throws Exception {
...
+        Iterables.size(unstablePartitioning.getVertices())
+            - Iterables.size(stablePartitioning.getVertices()));
+  }
+
+  private JobGraph getStatefulParDoAfterCombineChainingJobGraph(boolean stablePartitioning) {
+    final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    final FlinkStreamingPipelineTranslator translator =
+        new FlinkStreamingPipelineTranslator(env, PipelineOptionsFactory.create());
+    final PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    pipelineOptions.setRunner(FlinkRunner.class);
```
Comment (je-ik) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:158]: Do we need to set runner if we enforce the translator?

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java [hunk 8]
```diff
@@ -120,4 +138,95 @@ public void testAutoBalanceShardKeyCacheMaxSize() throws Exception {
...
+    final Pipeline pipeline = Pipeline.create(pipelineOptions);
+    PCollection<KV<String, Iterable<Long>>> aggregate =
+        pipeline
+            .apply(
+                Create.of(KV.of("foo", 1L), KV.of("bar", 1L))
+                    .withCoder(KvCoder.of(StringUtf8Coder.of(), VarLongCoder.of())))
+            .apply(GroupByKey.create());
```
Comment (je-ik) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:197]: Can we merge the two methods, that seem to differ only by this PTransform applied here?

**Review by je-ik — APPROVED** [review]
LGTM! A left a few comments, but feel free to merge this as it is.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java (hunk 7 above)
Comment (dmvk) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:158]: unfortunately yes, otherwise pipeline.create() would fail

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java (hunk 8 above)
Comment (dmvk) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslatorTest.java:197]: I thought about it, but the methods are really small so it's worth decreasing readability here

**Review by mxm — APPROVED** [review]
Thanks @dmvk. LGTM

**Comment by dmvk** [comment]
Run Flink ValidatesRunner
